nifi-nar-bundles/nifi-enrich-bundle/nifi-enrich-processors/src/main/java/org/apache/nifi/processors/GeoEnrichIPRecord.java

```diff
@@ -0,0 +1,323 @@
+        super.onScheduled(context);
+
+        readerFactory = context.getProperty(READER).asControllerService(RecordReaderFactory.class);
+        writerFactory = context.getProperty(WRITER).asControllerService(RecordSetWriterFactory.class);
+        splitOutput = context.getProperty(SPLIT_FOUND_NOT_FOUND).asBoolean();
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if (input == null) {
+            return;
+        }
+
+        FlowFile output = session.create(input);
+        FlowFile notFound = splitOutput ? session.create(input) : null;
+        final DatabaseReader dbReader = databaseReaderRef.get();
+        try (InputStream is = session.read(input);
+             OutputStream os = session.write(output);
+             OutputStream osNotFound = splitOutput ? session.write(notFound) : null) {
+            RecordPathCache cache = new RecordPathCache(GEO_PROPERTIES.size() + 1);
+            Map<PropertyDescriptor, RecordPath> paths = new HashMap<>();
+            for (PropertyDescriptor descriptor : GEO_PROPERTIES) {
+                if (!context.getProperty(descriptor).isSet()) {
+                    continue;
+                }
+                String rawPath = context.getProperty(descriptor).evaluateAttributeExpressions(input).getValue();
+                RecordPath compiled = cache.getCompiled(rawPath);
+                paths.put(descriptor, compiled);
+            }
+
+            String rawIpPath = context.getProperty(IP_RECORD_PATH).evaluateAttributeExpressions(input).getValue();
+            RecordPath ipPath = cache.getCompiled(rawIpPath);
+
+            RecordReader reader = readerFactory.createRecordReader(input, is, getLogger());
+            RecordSchema schema = writerFactory.getSchema(input.getAttributes(), null);
+            RecordSetWriter writer = writerFactory.createWriter(getLogger(), schema, os);
+            RecordSetWriter notFoundWriter = splitOutput ? writerFactory.createWriter(getLogger(), schema, osNotFound) : null;
+            Record record;
+            Relationship targetRelationship = REL_NOT_FOUND;
+            writer.beginRecordSet();
+
+            while ((record = reader.nextRecord()) != null) {
+                CityResponse response = geocode(ipPath, record, dbReader);
+                boolean wasEnriched = enrichRecord(response, record, paths);
+                if (wasEnriched) {
+                    targetRelationship = REL_FOUND;
+                }
+                if (!splitOutput || (splitOutput && wasEnriched)) {
+                    writer.write(record);
+                } else {
+                    notFoundWriter.write(record);
+                }
+            }
+            writer.finishRecordSet();
+            writer.close();
+            is.close();
+            os.close();
+            if (osNotFound != null) {
+                osNotFound.close();
+            }
+
```

Comment:
There are some other bookkeeping things that should be done for all record-based processors (for consistency), namely the setting of `mime.type` and `record.count` attributes, plus any others that might come from the writer. This is the standard code block (taken from `AbstractRouteRecord`):

```
final Map<String, String> attributes = new HashMap<>();
attributes.put("record.count", String.valueOf(writeResult.getRecordCount()));
attributes.put(CoreAttributes.MIME_TYPE.key(), writer.getMimeType());
attributes.putAll(writeResult.getAttributes());

childFlowFile = session.putAllAttributes(childFlowFile, attributes);
```
I found this while using a JSONRecordSetWriter, the MIME type came out as `text/plain` so I couldn't view the JSON in the UI.

Reply:
Ok. I'll add that, and I think we may need to add a separate Jira ticket for that JsonRecordSetWriter behavior.